lib/ecto/type.ex

```diff
@@ -714,19 +718,24 @@ defmodule Ecto.Type do
       :error
 
   """
-  @spec cast(t, term) :: {:ok, term} | {:error, keyword()} | :error
-  def cast({:parameterized, module, config}, value), do: module.cast(config, value)
-  def cast({:in, _type}, nil), do: :error
-  def cast(_type, nil), do: {:ok, nil}
+  @spec cast(t, term, term) :: {:ok, term} | {:error, keyword() | Ecto.Changeset.t()} | :error | :ignore
+  def cast(_, _, current \\ nil)
```

Comment:
So this is a bit complicated... `cast` is called both by `Ecto.Changeset` and also by `Ecto.Query`. Once we pass `current` here, we are lumping both of them together, but in one of them we will never actually have a `current`. To make matters further apart, an embed would never really cast for a query, only for a changeset.

So it may be that, for parameterized types, we will have two cast functions, one receives the current value and the other does not. We can also give different names to them. 

For `cast+current`, we don't need to add it to Ecto.Type, we can call it directly from the changeset. So in a nutshell, my suggestion is to remove the extra `current` argument from here for now. We will add a new `cast` function once we work on the changeset.

Reply:
Sounds good. Yeah, `current` came into play with `Embed`, but isn't necessary in v1.